### ISSUE TITLE
Bugs on DependencyMatcher

### ISSUE THREAD

**Issue by gnegrelli** [explosion/spaCy#6300]
Hello!

I was working on some patterns that require the `on_match` method of DependencyMatcher and found some bugs in it.

https://github.com/explosion/spaCy/blob/01aec7a313753775603a9e6f752f75cc16ac43fb/spacy/matcher/dependencymatcher.pyx#L236-L244

As it is today, the `on_match` method is executed after each match is added to the `matched_key_trees` instead of being called once after all patterns were evaluated. To fix it, the for loop in the snippet above should be outside the main for loop, so the on_match method is only executed once for each item in the `matched_key_trees`.

Also, the DependencyMatcher appends a tuple of `(key, matched_trees)` to the `matched_key_trees` even if there's no match (matched_trees is an empty list in this case). Shouldn't it only add in case `matched_trees` is not empty? A simple if statement would do?

## How to reproduce the behaviour

I wrote a simple test based on the script available in #4590 so you guys can have a look at it.

```python
from spacy.matcher import DependencyMatcher
from spacy import load


def my_on_match(matcher, doc, i, matches):
    print('inside my_on_match:')
    print('Matches found: {}'.format(matches))
    print('Treating match of \"{}\"'.format(doc.vocab.strings[matches[i][0]]))
    print('================')


nlp = load('en_core_web_sm')
doc = nlp('The quick brown fox jumped over the lazy dog.')

dep_matcher = DependencyMatcher(nlp.vocab)

dep_pattern_1 = [
    {"SPEC": {"NODE_NAME": "fox"}, "PATTERN": {"ORTH": "fox"}},
    {"SPEC": {"NODE_NAME": "quick", "NBOR_RELOP": ">", "NBOR_NAME": "fox"}, "PATTERN": {"ORTH": "quick"}},
]

dep_pattern_2 = [
    {"SPEC": {"NODE_NAME": "quick"}, "PATTERN": {"ORTH": "quick"}},
    {"SPEC": {"NODE_NAME": "brown", "NBOR_RELOP": "$+", "NBOR_NAME": "quick"}, "PATTERN": {"ORTH": "brown"}},
]

dep_pattern_3 = [
    {"SPEC": {"NODE_NAME": "quick"}, "PATTERN": {"ORTH": "quick"}},
    {"SPEC": {"NODE_NAME": "brown", "NBOR_RELOP": "$-", "NBOR_NAME": "quick"}, "PATTERN": {"ORTH": "brown"}},
]

dep_matcher.add('first_pattern', my_on_match, dep_pattern_1)
dep_matcher.add('second_pattern', my_on_match, dep_pattern_2)
dep_matcher.add('third_pattern', my_on_match, dep_pattern_3)

matches = dep_matcher(doc)

print('all matches')
for match in matches:
    print(match)
```

## Your Environment

* **spaCy version:** 2.1.8
* **Platform:** Linux-5.4.0-51-generic-x86_64-with-Ubuntu-18.04-bionic
* **Python version:** 3.6.9

**Comment by adrianeboyd** [comment]
Thanks for the report! Those do both look like bugs. If you'd like to submit a PR, we'd be happy to accept it, but it's also no problem if you'd prefer for us to handle it.

**Comment by gnegrelli** [comment]
Hi, @adrianeboyd. 

Sorry I didn't answer your question. :disappointed: Anyway, thanks for the PR!

**Comment by adrianeboyd** [comment]
Do you need this fix in v2.1 in particular? We can backport bug fixes on request, but the current PR would only end up in v2.3.3 as it stands.

**Comment by gnegrelli** [comment]
Yeah! If you could also backport it to v2.2 as well, that would be great!
